erdjs/src/dapp/walletProvider.ts

```diff
@@ -0,0 +1,237 @@
+    DAPP_MESSAGE_INIT,
+    DAPP_DEFAULT_TIMEOUT,
+    DAPP_MESSAGE_IS_CONNECTED,
+    DAPP_MESSAGE_CONNECT_URL, DAPP_MESSAGE_SEND_TRANSACTION_URL
+} from "./constants";
+import {mainFrameStyle} from "./dom";
+import {Transaction} from "../transaction";
+
+export class WalletProvider implements IDappProvider {
+    walletUrl: string;
+    mainFrame: HTMLIFrameElement | undefined;
+
+    /**
+     * Creates a new WalletProvider
+     * @param walletURL
+     */
+    constructor(walletURL: string = '') {
+        this.walletUrl = walletURL;
+        this.attachMainFrame();
+    }
+
+    /**
+     * Waits for the wallet iframe to ping that it has been initialised
+     */
+    async init(): Promise<boolean> {
+        if (this.isInitialized()) {
+            return true;
+        }
+        return this.waitForRemote();
+    }
+
+    /**
+     * Returns if the wallet iframe is up and running
+     */
+    isInitialized(): boolean {
+        if (!this.mainFrame) {
+            return false;
+        }
+
+        return this.mainFrame.dataset.initialized === "true";
+    }
+
+    /**
+     * Unlike isInitialized, isConnected returns true if the user alredy went through the login process
+     *  and has the wallet session active
+     */
+    async isConnected(): Promise<boolean> {
+        if (!this.mainFrame) {
+            return false;
+        }
+
+        const {contentWindow} = this.mainFrame;
+        if (!contentWindow) {
+            return false;
+        }
+
+        return new Promise((resolve, reject) => {
+            contentWindow.postMessage({
+                type: DAPP_MESSAGE_IS_CONNECTED,
+            }, this.walletUrl);
+
+            const timeout = setTimeout(_ => reject('window not responding'), 5000);
+
+            const isConnected = (ev: IDappMessageEvent) => {
+                if (!this.isValidWalletSource(ev.origin)) {
+                    return;
+                }
+
+                const {data} = ev;
+                if (data.type !== DAPP_MESSAGE_IS_CONNECTED) {
+                    return;
+                }
+
+                clearTimeout(timeout);
+                window.removeEventListener('message', isConnected.bind(this));
+                return resolve(data.data);
+            };
+
+            window.addEventListener('message', isConnected);
+        });
+    }
+
+    /**
+     * Fetches the login hook url and redirects the client to the wallet login.
+     */
+    async login(): Promise<string> {
+        if (!this.mainFrame) {
+            return '';
+        }
+
+        const {contentWindow} = this.mainFrame;
+        if (!contentWindow) {
+            return '';
+        }
+
+        return new Promise<string>((resolve, reject) => {
+            contentWindow.postMessage({
+                type: DAPP_MESSAGE_CONNECT_URL,
+            }, this.walletUrl);
+
+            const timeout = setTimeout(_ => reject('connect url not responding'), 5000);
+            const connectUrl = (ev: IDappMessageEvent) => {
```

Comment:
Can we extract `connectUrl` to a separate, private function?

Reply:
I've done it like this because this callbacks need access to the local `timeout`. If you'd like I can think of how it can be refactored